python/pyspark/mllib/linalg.py

```diff
@@ -1152,9 +1156,385 @@ def sparse(numRows, numCols, colPtrs, rowIndices, values):
+        ...                        IndexedRow(2, Vectors.dense([7, 8, 9])),
+        ...                        IndexedRow(3, Vectors.dense([10, 11, 12]))])
+        >>> rm = IndexedRowMatrix(rows, 7, 6)
+        >>> int(rm.numCols())
+        6
+        """
+        return self._jirm.call("numCols")
+
+    def toRowMatrix(self):
+        """
+        Convert this matrix to a RowMatrix.
+
+        >>> rows = sc.parallelize([IndexedRow(0, Vectors.dense([1, 2, 3])),
+        ...                        IndexedRow(6, Vectors.dense([4, 5, 6]))])
+        >>> rm = IndexedRowMatrix(rows).toRowMatrix()
+        >>> rm.rows.collect()
+        [DenseVector([1.0, 2.0, 3.0]), DenseVector([4.0, 5.0, 6.0])]
+        """
+        javaRowMatrix = self._jirm.call("toRowMatrix")
+        return RowMatrix._from_java(javaRowMatrix)
+
+    def toCoordinateMatrix(self):
+        """
+        Convert this matrix to a CoordinateMatrix.
+
+        >>> rows = sc.parallelize([IndexedRow(0, Vectors.dense([1, 0])),
+        ...                        IndexedRow(6, Vectors.dense([0, 5]))])
+        >>> cm = IndexedRowMatrix(rows).toCoordinateMatrix()
+        >>> cm.entries.take(3)
+        [MatrixEntry(0, 0, 1.0), MatrixEntry(0, 1, 0.0), MatrixEntry(6, 0, 0.0)]
+        """
+        javaCoordinateMatrix = self._jirm.call("toCoordinateMatrix")
+        return CoordinateMatrix._from_java(javaCoordinateMatrix)
+
+
+class MatrixEntry(object):
+    """
+    Represents an entry of a CoordinateMatrix.
+
+    Just a wrapper over a (long, long, float) tuple.
+
+    .. note:: Experimental
+    """
+    def __init__(self, i, j, value):
+        self.i = long(i)
+        self.j = long(j)
+        self.value = float(value)
+
+    def __repr__(self):
+        return "MatrixEntry(%s, %s, %s)" % (self.i, self.j, self.value)
+
+
+class CoordinateMatrix(object):
+    """
+    Represents a matrix in coordinate format.
+
+    .. note:: Experimental
+    """
+    def __init__(self, entries, numRows=0, numCols=0):
+        """
+        Create a wrapper over a Java CoordinateMatrix.
+
+        :param entries: An RDD of MatrixEntry inputs or
+                        (long, long, float) tuples.
+        """
+        if not isinstance(entries, RDD):
+            raise TypeError("entries should be an RDD object, got %s" % type(entries))
+        first = entries.first()
+        if isinstance(first, MatrixEntry):
+            pass
+        elif (isinstance(first, (tuple, list)) and len(first) == 3 and
+              isinstance(first[0], (long, int)) and
+              isinstance(first[1], (long, int)) and
+              isinstance(first[2], (float, long, int))):
+            # The Java side won't be able to automatically convert a
+            # long to a double for the 3rd tuple value, so map each
+            # tuple to a MatrixEntry, which will do the conversion
+            # on the Python side.
+            entries = entries.map(lambda entry: MatrixEntry(*entry))
```

Comment:
It seems that there is already typecasting in MatrixEntry as well. You could do away with the checks above.